hs/src/Reach/Simulator/Core.hs

```diff
@@ -660,70 +648,127 @@ instance Interp LLStep where
       _ <- interp stmt
       interp step
     LLS_Stop _at -> return V_Null
-    LLS_ToConsensus _at _lct tc_send recv tc_mtime -> do
+    LLS_ToConsensus _at _lct tc_send dlr@(DLRecv {..}) tc_mtime -> do
       (g, l) <- getState
-      let mid = e_nmsgid g
-      incrMessageId
+      let actId = l_curr_actor_id l
+      phId <- getPhaseId actId
+      msgs' <- case M.lookup phId $ e_messages g of
+            Just m -> return m
+            Nothing -> do
+              let msg = NotFixedYet mempty
+              setGlobal g { e_messages = M.insert phId msg $ e_messages g }
+              return msg
       let sends = M.mapKeys bunpack tc_send
-      v <- suspend $ PS_Suspend (A_TieBreak mid $ M.keys sends)
-      case v of
-        V_UInt n -> do
-          let locals = e_locals l
-          let lclsv = saferMapRef "LLS_ToConsensus" $ M.lookup (fromIntegral n) locals
-          let lclst = l_who $ lclsv
-          case lclst of
-            Nothing -> impossible "expected participant id, received consensus id"
-            Just part -> do
+      incrPhaseId
+      isTheTimePast tc_mtime >>= \case
+        Just sk -> interp sk
+        Nothing -> do
+          whoAmI >>= \case
+            Participant who -> do
+              case M.lookup who sends of
+                Nothing -> do
+                  case msgs' of
+                    NotFixedYet _msgs'' -> do
+                      _ <- suspend $ PS_Suspend (A_Contest phId)
+                      winner dlr actId phId
+                      interp $ dr_k
+                    Fixed (actId',_msg) -> do
+                      winner dlr actId' phId
+                      interp $ dr_k
+                Just (DLSend {..}) -> do
+                  case msgs' of
+                    NotFixedYet msgs'' -> do
+                      ds_msg' <- mapM interp ds_msg
+                      let sto = M.fromList $ zip dr_msg ds_msg'
+                      setGlobal g { e_messages = M.insert phId (NotFixedYet $ M.insert actId sto msgs'') (e_messages g) }
```

Comment:
Because you're just going to be setting it down hear anyways

Reply:
fair